File: 0921-minimum-add-to-make-parentheses-valid/0921-minimum-add-to-make-parentheses-valid.java
class Solution {
    public int minAddToMakeValid(String s) {
        Stack<Character> st = new Stack<>();
        
        for(int i = 0;i<s.length();i++){
            char ch = s.charAt(i);
            
            // if we have opening bracket 
            if(ch == '('){
                st.push(ch);
            }
            // if we have closing bracket then we have to check for the peek of the stack for the opening bracket. 
            else{
                if(st.size() > 0 && st.peek() == '('){
                    st.pop();
                }
                else{
                    st.push(ch);
                }
            }
        }
        return st.size();
    }
}

File: 0921-minimum-add-to-make-parentheses-valid/0921-minimum-add-to-make-parentheses-valid.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
         stack<char> st;
        
        for(int i = 0;i<s.size();i++){
            if(s[i] == '('){
                st.push(s[i]);
            }
            else{
                //check for the top element of the stack.
                if(st.size() > 0 && st.top() == '('){
                    st.pop();
                }
                else{
                    st.push(')');
                }
            }
        }
        return st.size();
    }
};